### PR TITLE
add dtc-package generating dtc, libfdt and pylibfdt

### PR DESCRIPTION
**Pull request by timsnyder**
mostly copied from LiteX-Hub's recipe but merged in conda-forgeness
* split into 3 packages dtc, libfdt, pylibfdt.
   * licenses are different between the three
   * minmize impact of python linkage & dep
* added optional yaml dependency to dtc

Checklist
- [X] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [X] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [X] Source is from official source.
- [X] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [X] If static libraries are linked in, the license of the static library is packaged.
- [X] Build number is 0.
- [X] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [X] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [X] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/dtc/meta.yaml
Comment (timsnyder) [inline recipes/dtc/meta.yaml:87]:
```suggestion
    script: install.sh   # [not win]
    script: install.bat  # [win]
```

File: recipes/dtc/meta.yaml
Comment (timsnyder) [inline recipes/dtc/meta.yaml:99]:
```suggestion
        - test -f $PREFIX/lib/libfdt-{{ version }}${SHLIB_EXT}  # [not win]
```

**Comment by timsnyder** [comment]
Seems like the linter is complaining about there not being a top-level about when all of the split packages have their own about sections.  This is already captured in conda-smithy#1124 by @jakirkham and I'll just do what he ended up doing in #9162 and add the toplevel about section since I know it won't be used.

**Comment by timsnyder** [comment]
Since dtc is a very linux-specific thing, I was thinking it might only make sense on linux but the fact that flex and bison aren't available on windows makes it definitely not buildable on Windows.  Will set it up to only build on unixen.

File: recipes/dtc/meta.yaml
Comment (timsnyder) [inline recipes/dtc/meta.yaml:114]:
```suggestion
  license: GPL-2.0-or-later
```

File: recipes/dtc/meta.yaml
Comment (timsnyder) [inline recipes/dtc/meta.yaml:114]:
```suggestion
  license: GPL-2.0-or-later
  license_family: GPL
  license_file:
    - GPL
    - README.license
```

File: recipes/dtc/conda-forge.yml
Comment (timsnyder) [inline recipes/dtc/conda-forge.yml:2]: I couldn't remember whether I need to do this manually here or if it happens because I have `build.skip: True  # [win]` in my meta.yaml once we convert this into a feedstock.  I'm beginning to think I should delete this file before we merge this.

File: recipes/dtc/meta.yaml
Comment (timsnyder) [inline recipes/dtc/meta.yaml:37]: I also don't remember having to specify `/lib64/libc.so.6` in the missing_dso_whitelist before.  I searched the other feedstocks and found several that do this.  Does this mean that I'm directly linking to the system libc instead of the one in the compiler sysroot?  I'm rereading https://github.com/conda/conda-build/blob/0a6efdcfa02484954e7a3fdf5c2ec4bb3cb6682e/conda_build/post.py.  It's been awhile.

Comment (timsnyder) [reply recipes/dtc/meta.yaml:37]: Looking through the other recipes for CONDA_BUILD_SYSROOT, I see a bunch of packages adding `-isysroot $CONDA_BUILD_SYSROOT` into CFLAGS CXXFLAGS and LDFLAGS.  I tried doing that and removing the missing_dso_whitelist and I get the errors again.  They look like this:
```
Warning: rpath /home/conda/staged-recipes/build_artifacts/dtc-package_1645834285171/_build_env/lib is outside prefix /home/conda/staged-recipes/build_artifacts/dtc-package_1645834285171/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold (removing it) for /home/conda/staged-recipes/build_artifacts/dtc-package_1645834285171/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/libfdt-1.6.1.so
WARNING :: Failed to get_static_lib_exports(/home/conda/staged-recipes/build_artifacts/dtc-package_1645834285171/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/libfdt.a)
Unknown format
  ERROR (libfdt,lib/libfdt-1.6.1.so): /lib64/libc.so.6 not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
```

@conda-forge/help-c-cpp what's the deal with libc and the missing_dso_whitelist?  At one point, I thought that we expected stuff to link against the sysroot libc at build time and then at runtime link the system libc.  Should I just leave the missing_dso_whitelist in my recipe or do I need to modify the build so that it links differently?

Comment (timsnyder) [reply recipes/dtc/meta.yaml:37]: I think I got this figured out after looking more at https://github.com/conda/conda-build/blob/77463386355f3a726437311d3fa1f39a59afc2e9/conda_build/post.py#L1062-L1072, I realized that the split packages need to list the compiler in the `requirements.build` section of their output so that the sysroot will be found in build_prefix and libc will be found to be okay by the overlinking logic.  Testing this locally and will push an update to remove the missing_dso_whitelist and also remove the `-isysroot` arguments because I'm pretty sure that those are baked into the compilers correctly for linux.  Perhaps they are needed for osx though?

**Comment by timsnyder** [comment]
@conda-forge/staged-recipes @conda-forge/help-c-cpp this one is ready to merge if I can get a review.  Thanks.

File: recipes/dtc/meta.yaml
Comment (carterbox) [inline recipes/dtc/meta.yaml:67]: pylibfdt statically links to libfdt? Is there a compile option to link dynamically?

File: recipes/dtc/meta.yaml
Comment (carterbox) [inline recipes/dtc/meta.yaml:46]: What about fdtget, fdtdump, etc.? Can you not test those binaries too?

File: recipes/dtc/meta.yaml
Comment (carterbox) [inline recipes/dtc/meta.yaml:94]: There are two additional headers being shipped.

Test for .so .so.1 .a too!

**Review by carterbox — CHANGES_REQUESTED** [review]
Hi @timsnyder,

You can check the outputs of these builds by downloading the build artifacts from azure. https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=465595&view=artifacts&pathAsName=false&type=publishedArtifacts

I noticed that there were a few more things being shipped compared to what you tested for, so double check that you're only shipping what you expect to be included!

File: recipes/dtc/meta.yaml
Comment (timsnyder) [reply recipes/dtc/meta.yaml:67]: The extension links against `libfdt.a` when creating (for example) `lib/python3.8/site-packages/_libfdt.cpython-38-x86_64-linux-gnu.so`.  So, the pylibfdt package, does not depend at runtime on the libfdt package.  I don't see a way to change the linkage but I'm not an expert when it comes to building python extensions.

File: recipes/dtc/meta.yaml
Comment (timsnyder) [reply recipes/dtc/meta.yaml:46]: Sure.  I'll add more tests.

File: recipes/dtc/meta.yaml
Comment (timsnyder) [reply recipes/dtc/meta.yaml]: Done

**Comment by timsnyder** [comment]
@carterbox Thanks for the review.  I added more tests and assuming that I didn't break the osx build, I think this should be ready to merge if you can give it another pass.

File: recipes/dtc/meta.yaml
Comment (carterbox) [inline recipes/dtc/meta.yaml:79]: Could you explain what this test is for? It looks like you already tested all of the binaries in previous tests (either by running help or checking for their existence).

File: recipes/dtc/meta.yaml
Comment (carterbox) [reply recipes/dtc/meta.yaml:100]: Lemme look into this whilst you fix the broken OSX build. Otherwise, we'll just ship it statically linked.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by timsnyder** [comment]
'make check' is running the actual test suite in the package to make sure
that more functionality works in binaries and that the libraries can be
linked against once installed.

We could probably remove the existence tests now but I left them in since
they aren't hurting anything.

**Comment by timsnyder** [comment]
I'll need to look into the osx failure later this evening.

**Comment by timsnyder** [comment]
In dtc, `tests/trees.S` uses GCC-specific assembly directives  to build up test collateral binary files into a utility called dumptrees and then spit them out into various .dtb files that the tests manipulate.  I looked briefly into getting that to build using clang and I don't think it will be trivial.  I'm going to do the additional testing only on Linux where it is known to work and only do existence and import testing on osx.

**Comment by timsnyder** [comment]
@carterbox any luck with figuring out how to have the python package use the dynamic library from libfdt?  I'm poking around trying a few things this morning.

**Comment by timsnyder** [comment]
I don't see a way to make the python module use the dynamic libfdt.  @carterbox can we just ship the statically linked one?

**Comment by timsnyder** [comment]
@carterbox , I saw an email with a big comment that I guess you deleted about how pylibfdt is being built.  It looks like the upstream folks that work on dtc are also noticing and working on some fixes to get pylibfdt to depend on libfdt in https://www.spinics.net/lists/devicetree-compiler/msg03891.html from about a month ago.  I'll keep an eye on that and update the recipe when the next version is cut that includes these improvements (hopefully).

**Comment by carterbox** [comment]
@timsnyder, not deleted, just [buried in a thread](https://github.com/conda-forge/staged-recipes/pull/18166#discussion_r821911058). 😄

> I'll keep an eye on that and update the recipe when the next version 

Sounds good.

**Pull request merged**
